text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+```js
+person.fullName; // 'Rob Dale'
+
+person.lastName = 'Jackson';
+```
+
+Now `fullName` knows which tracked properties were accessed when it was run,
+and setting `lastName` has invalidated `fullName`.
+
+> **NOTE:** This does _not_ invalidate a cache like in computed properties. Even
+> if `firstName` and `lastName` were untracked, the tracked getter would still
+> return the correct value on subsequent accesses, because `@tracked` does _not_
+> cache values. The validation and invalidation is pure meta data that is only
+> accessible by the Glimmer VM.
+>
+> Internally, Glimmer checks to see if a value has updated _before calling the
+> getter_. If it hasn't, then Glimmer does not rerender the related section of
+> the DOM. This is effectively an automatic `shouldComponentUpdate` from React.
+>
+> To prevent inconsistency, during development time, tracked properties will
+> keep a cache of their previous value to compare when they are activated and
+> ensure that it hasn't changed without invalidation. This will prevent improper
+> usage of tracked properties _outside_ of Glimmer's change tracking.
+
+### Manual Invalidation
+
+In user code, the idea that all mutable properties should be marked as tracked
+and that all other properties are effectively immutable works well in isolation.
+However, there are cases where users will want to work with code they do _not_
+control, such as external library code.
+
+Consider the following example. We have a `simple-timer` library that we've
+imported from NPM, and we're trying to wrap it with a `TimerComponent` that
+uses it to keep track of how much time has passed:
+
+```js
+// simple-timer/index.js
+export default class Timer {
+  seconds = 0;
+  minutes = 0;
+  hours = 0;
+
+  listeners = [];
+
+  constructor() {
+    setInterval(() => {
+      this.seconds++;
+      this.minutes = Math.floor(this.seconds / 60);
+      this.hours = Math.floor(this.minutes / 60);
+      this.notifyTick();
+    }, 1000);
+  }
+
+  notifyTick() {
+    for (let listener of this.listeners) {
+      listener(this.seconds);
+    }
+  }
+
+  onTick(listener) {
+    this.listeners.push(listener);
+  }
+}
+```
+
+```js
+import Timer from 'simple-timer';
+import Component, { tracked } from '@glimmer/tracking';
+
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  @tracked
+  get currentSeconds() {
+    return this.timer.seconds;
+  }
+
+  @tracked
+  get currentMinutes() {
+    return this.timer.minutes;
+  }
+}
+```
+
+Even though we've marked the `timer` property as tracked, the `timer.seconds`
+property is untracked, and _it_ is the field that is updated. We can solve this
+problem by using the timer library's `onTick` event handler to re-set the field,
+invalidating it:
+
+```js
+export default class TimerComponent extends Component {
+  @tracked timer = new Timer();
+
+  constructor() {
+    this.timer.onTick(() => {
+      // invalidate the timer field.
```

Comment:
if I came across this in code, I'd be confused.
I think the comment should be a bit more descriptive, like:
```
// in order to bust the `@tracked` "cache", we need to set the timer, so 
// anything depending on timer for display or computing can access the
// updated values from within the timer object.
```

Reply:
This is related to the above thread about adding `invalidate()`/`notify()`, going to mark this as resolved and we can continue the discussion there.